privaterelay/utils.py

```diff
@@ -27,34 +40,242 @@ def get_countries_info_from_request_and_mapping(request, mapping):
+    "trs": "mx",  # Triqui -> Mexico
+    "uk": "ua",  # Ukrainian -> Ukraine
+    "ur": "pk",  # Urdu -> Pakistan
+    "uz": "uz",  # Uzbek -> Uzbekistan
+    "vi": "vn",  # Vietnamese -> Vietnam
+    "wo": "sn",  # Wolof -> Senegal
+    "xcl": "am",  # Armenian Classic -> Armenia
+    "xh": "za",  # Xhosa -> South Africa
+    "zam": "mx",  # Miahuatlán Zapotec -> Mexico
+    "zh": "cn",  # Chinese -> China
+}
+
+# Special cases for language tags
+_LANGUAGE_TAG_TO_COUNTRY_OVERRIDE = {
+    # Would be Catalan in Valencian script -> France
+    # Change to Valencian -> Spain
+    ("ca", "valencia"): "es",
+    # Spanish in UN region 419 (Latin America and Carribean)
+    # Pick Mexico, which has highest number of Spanish speakers
+    ("es", "419"): "mx",
+    # Would be Galician (Greenland) -> Greenland
+    # Change to Galician (Galicia region of Spain) -> Spain
+    ("gl", "gl"): "es",
+}
+
+
+class AcceptLanguageError(ValueError):
+    """There was an issue processing the Accept-Language header."""
+
+    def __init__(self, message, accept_lang):
+        super().__init__(message)
+        self.accept_lang = accept_lang
+
+
+def guess_country_from_accept_lang(accept_lang: str) -> str:
+    """
+    Guess the user's country from the Accept-Language header
+
+    Return is a 2-letter ISO 3166 country code, lowercased.
+
+    If an issue is detected, a AcceptLanguageError is raised.
+
+    The header may come directly from a web request, or may be the header
+    captured by Firefox Accounts (FxA) at signup.
+
+    Even with all this logic and special casing, it is still more accurate to
+    use a GeoIP lookup or a country code provided by the infrastructure.
+
+    See RFC 9110, "HTTP Semantics", section 12.5.4, "Accept-Language"
+    See RFC 5646, "Tags for Identifying Languages", and examples in Appendix A
+    """
+    lang_q_pairs = parse_accept_lang_header(accept_lang.strip())
+    if not lang_q_pairs:
+        raise AcceptLanguageError("Invalid Accept-Language string", accept_lang)
+    top_lang_tag = lang_q_pairs[0][0]
+
+    subtags = top_lang_tag.split("-")
+    lang = subtags[0].lower()
+    if lang == "i":
+        raise AcceptLanguageError("Irregular language tag", accept_lang)
+    if lang == "x":
+        raise AcceptLanguageError("Private-use language tag", accept_lang)
+    if lang == "*":
+        raise AcceptLanguageError("Wildcard language tag", accept_lang)
+    if len(lang) < 2:
+        raise AcceptLanguageError("Invalid one-character primary language", accept_lang)
+    if len(lang) == 3 and lang[0] == "q" and lang[1] <= "t":
+        raise AcceptLanguageError(
+            "Private-use language tag (RFC 5646 2.2.1)", accept_lang
+        )
+
+    for maybe_region_raw in subtags[1:]:
+        maybe_region = maybe_region_raw.lower()
+
+        # Look for a special case
+        try:
+            return _LANGUAGE_TAG_TO_COUNTRY_OVERRIDE[(lang, maybe_region)]
+        except KeyError:
+            pass
```

Comment:
comment (non-blocking): hmm ... I'm not sure why but this `try ... except KeyError: pass` logic feels looser than:
```python
if (lang, maybe_region) in _LANGUAGE_TAG_TO_COUNTRY_OVERRIDE:
    return _LANGUAGE_TAG_TO_COUNTRY_OVERRIDE[(lang, maybe_region)]
```

I suppose doing it this way reduces all that code duplication, and this code is well tested. I just don't think I've ever seen this approach before.

Reply:
It's a long debate, here's an argument from 2016:

https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/

After re-reading it, it is much more common for the override to NOT be present, so Brett would agree with you that I picked the wrong one. I'm looking for ways to use the walrus operator, is this too ugly?

```suggestion
        if (override := _LANGUAGE_TAG_TO_COUNTRY_OVERRIDE.get((lang, maybe_region))):
            return override
```

Reply:
TIL (Today I loved): the walrus operator.